### PR TITLE
`lit_to_const`: gracefully bubble up type errors.

### LINKED ISSUE #69310: internal compiler error: src/librustc_mir_build/hair/constant.rs:60: impossible case reached
I'm seeing an internal compiler error on the following input (found by [fuzz-rustc](https://github.com/dwrensha/fuzz-rustc)):

```rust
const X: [(); 0.1] = [()];
fn main() {}
```

```
$ rustc main.rs
error: internal compiler error: src/librustc_mir_build/hair/constant.rs:60: impossible case reached

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:881:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.43.0-nightly (7760cd0fb 2020-02-19) running on x86_64-unknown-linux-gnu

error: aborting due to previous error
```

The error happens on `nightly` and `beta`, but not on `stable`.

<details>
<summary>Nightly backtrace</summary>

```

error: internal compiler error: src/librustc_mir_build/hair/constant.rs:60: impossible case reached

thread 'rustc' panicked at 'Box<Any>', src/librustc_errors/lib.rs:881:9
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.44/src/backtrace/libunwind.rs:86
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.44/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:78
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1052
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1428
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:204
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:224
  10: rustc_driver::report_ice
  11: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:474
  12: std::panicking::begin_panic
  13: rustc_errors::HandlerInner::bug
  14: rustc_errors::Handler::bug
  15: rustc::util::bug::opt_span_bug_fmt::{{closure}}
  16: rustc::ty::context::tls::with_opt::{{closure}}
  17: rustc::ty::context::tls::with_opt
  18: rustc::util::bug::opt_span_bug_fmt
  19: rustc::util::bug::bug_fmt
  20: rustc_mir_build::hair::constant::lit_to_const
  21: rustc::ty::query::__query_compute::lit_to_const
  22: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::lit_to_const>::compute
  23: rustc::dep_graph::graph::DepGraph::with_task_impl
  24: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  25: <dyn rustc_typeck::astconv::AstConv>::ast_const_to_const
  26: <dyn rustc_typeck::astconv::AstConv>::ast_ty_to_ty
  27: rustc_typeck::collect::type_of::type_of
  28: rustc::ty::query::__query_compute::type_of
  29: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::type_of>::compute
  30: rustc::dep_graph::graph::DepGraph::with_task_impl
  31: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  32: <rustc_typeck::collect::CollectItemTypesVisitor as rustc_hir::intravisit::Visitor>::visit_item
  33: rustc::hir::map::Map::visit_item_likes_in_module
  34: rustc_typeck::collect::collect_mod_item_types
  35: rustc::ty::query::__query_compute::collect_mod_item_types
  36: rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors for rustc::ty::query::queries::collect_mod_item_types>::compute
  37: rustc::dep_graph::graph::DepGraph::with_task_impl
  38: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  39: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::ensure_query
  40: rustc_session::session::Session::track_errors
  41: rustc_typeck::check_crate
  42: rustc_interface::passes::analysis
  43: rustc::ty::query::__query_compute::analysis
  44: rustc::dep_graph::graph::DepGraph::with_task_impl
  45: rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt>::get_query
  46: rustc::ty::context::tls::enter_global
  47: rustc_interface::interface::run_compiler_in_existing_thread_pool
  48: scoped_tls::ScopedKey<T>::set
  49: syntax::attr::with_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.43.0-nightly (7760cd0fb 2020-02-19) running on x86_64-unknown-linux-gnu

note: compiler flags: -C codegen-units=1 -C debuginfo=2 --crate-type bin

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [lit_to_const] converting literal to const
#1 [type_of] processing `X`
#2 [collect_mod_item_types] collecting item types in top-level module
#3 [analysis] running analysis passes on this crate
end of query stack
```

</details>

### PR DESCRIPTION
**Pull request by Centril**
Fixes https://github.com/rust-lang/rust/issues/69310 which was injected by https://github.com/rust-lang/rust/pull/68118.

r? @pnkfelix @varkor @eddyb 
cc @skinny121

### CODE REVIEW

File: src/librustc_mir_build/hair/constant.rs [hunk 1]
```diff
@@ -20,50 +20,35 @@ crate fn lit_to_const<'tcx>(
...
-                        bug!("bytestring should have type of either &[u8] or &[u8; _], not {}", ty)
-                    }
-                }
-            } else {
-                bug!("bytestring should have type of either &[u8] or &[u8; _], not {}", ty)
-            }
+        (ast::LitKind::ByteStr(data), ty::Ref(_, TyS { kind: ty::Slice(_), .. }, _)) => {
+            let allocation = Allocation::from_byte_aligned_bytes(data as &Vec<u8>);
+            let allocation = tcx.intern_const_alloc(allocation);
+            ConstValue::Slice { data: allocation, start: 0, end: data.len() }
```
Comment (eddyb) [inline src/librustc_mir_build/hair/constant.rs]: Reminder to self that reviewing this is best done in "No Whitespace" mode.

Comment (Centril) [reply src/librustc_mir_build/hair/constant.rs]: Indeed :)

File: src/librustc_mir_build/hair/cx/mod.rs [hunk 2]
```diff
@@ -148,6 +148,7 @@ impl<'a, 'tcx> Cx<'a, 'tcx> {
                 // create a dummy value and continue compiling
                 Const::from_bits(self.tcx, 0, self.param_env.and(ty))
```
Comment (eddyb) [inline src/librustc_mir_build/hair/cx/mod.rs]: @oli-obk This isn't new in this PR but I wonder why this isn't undef or something (I guess that might cause ICEs?). Maybe we need a `ty::ConstKind::Err` (if we don't have it already)?

File: src/librustc/mir/interpret/mod.rs [hunk 3]
```diff
@@ -148,6 +148,8 @@ pub struct LitToConstInput<'tcx> {
 /// Error type for `tcx.lit_to_const`.
 #[derive(Copy, Clone, Debug, Eq, PartialEq, HashStable)]
 pub enum LitToConstError {
+    /// The literal's inferred type did not match the expected `ty` in the input.
+    TypeError,
```
Comment (eddyb) [inline src/librustc/mir/interpret/mod.rs]: Maybe leave a comment that this only needs graceful handling in type-checking (which I didn't even realize was using this!), and everything else can ICE on it.

File: src/librustc_mir_build/hair/cx/mod.rs (hunk 2 above)
Comment (Centril) [reply src/librustc_mir_build/hair/cx/mod.rs]: We don't have such an `::Err` variant, https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/enum.ConstKind.html.

File: src/librustc_mir_build/hair/constant.rs [hunk 4]
```diff
@@ -20,50 +20,35 @@ crate fn lit_to_const<'tcx>(
...
+        (ast::LitKind::Float(n, _), ty::Float(fty)) => {
+            parse_float(*n, *fty, neg).map_err(|_| LitToConstError::UnparseableFloat)?
         }
-        ast::LitKind::Bool(b) => ConstValue::Scalar(Scalar::from_bool(b)),
-        ast::LitKind::Char(c) => ConstValue::Scalar(Scalar::from_char(c)),
-        ast::LitKind::Err(_) => return Err(LitToConstError::Reported),
+        (ast::LitKind::Bool(b), ty::Bool) => ConstValue::Scalar(Scalar::from_bool(*b)),
+        (ast::LitKind::Char(c), ty::Char) => ConstValue::Scalar(Scalar::from_char(*c)),
+        (ast::LitKind::Err(_), _) => return Err(LitToConstError::Reported),
+        (_, _) => return Err(LitToConstError::TypeError),
```
Comment (eddyb) [inline src/librustc_mir_build/hair/constant.rs]: This can also be `_ => `, but I guess it doesn't matter.

**Review by eddyb — APPROVED** [review]
r=me modulo nits

File: src/librustc_mir_build/hair/cx/mod.rs (hunk 2 above)
Comment (eddyb) [reply src/librustc_mir_build/hair/cx/mod.rs]: Adding one is probably a good idea (but can be done in a separate PR).

**Comment by Centril** [comment]
Nits fixed, r? @eddyb @bors r=eddyb

**Comment by pnkfelix** [comment]
discussed in T-compiler meeting. beta-accepted.

**Pull request merged**